src/xsdba/utils.py

```diff
@@ -0,0 +1,1238 @@
+        - `end` : Last index in the cluster, inclusive (`dim` reduced, new `cluster`), int
+        - `maxpos` : Index of the maximal value within the cluster (`dim` reduced, new `cluster`), int
+        - `maximum` : Maximal value within the cluster (`dim` reduced, new `cluster`), same dtype as data.
+
+      For `start`, `end` and `maxpos`, -1 means nan and should always correspond to a `nan` in `maximum`.
+      The length along `cluster` is half the size of "dim", the maximal theoretical number of clusters.
+    """
+
+    def _get_clusters(arr, u1, u2, N):
+        st, ed, mp, mv = get_clusters_1d(arr, u1, u2)
+        count = len(st)
+        pad = [-1] * (N - count)
+        return (
+            np.append(st, pad),
+            np.append(ed, pad),
+            np.append(mp, pad),
+            np.append(mv, [np.nan] * (N - count)),
+            count,
+        )
+
+    # The largest possible number of clusters. Ex: odd positions are < u2, even positions are > u1.
+    N = data[dim].size // 2
+
+    starts, ends, maxpos, maxval, nclusters = xr.apply_ufunc(
+        _get_clusters,
+        data,
+        u1,
+        u2,
+        input_core_dims=[[dim], [], []],
+        output_core_dims=[["cluster"], ["cluster"], ["cluster"], ["cluster"], []],
+        kwargs={"N": N},
+        dask="parallelized",
+        vectorize=True,
+        dask_gufunc_kwargs={
+            "meta": (
+                np.array((), dtype=int),
+                np.array((), dtype=int),
+                np.array((), dtype=int),
+                np.array((), dtype=data.dtype),
+                np.array((), dtype=int),
+            ),
+            "output_sizes": {"cluster": N},
+        },
+    )
+
+    ds = xr.Dataset(
+        {
+            "start": starts,
+            "end": ends,
+            "maxpos": maxpos,
+            "maximum": maxval,
+            "nclusters": nclusters,
+        }
+    )
+
+    return ds
+
+
+def rand_rot_matrix(
+    crd: xr.DataArray, num: int = 1, new_dim: str | None = None
+) -> xr.DataArray:
+    r"""Generate random rotation matrices.
+
+    Rotation matrices are members of the SO(n) group, where n is the matrix size (`crd.size`).
+    They can be characterized as orthogonal matrices with determinant 1. A square matrix :math:`R`
+    is a rotation matrix if and only if :math:`R^t = R^{−1}` and :math:`\mathrm{det} R = 1`.
+
+    Parameters
+    ----------
+    crd: xr.DataArray
+      1D coordinate DataArray along which the rotation occurs.
+      The output will be square with the same coordinate replicated,
+      the second renamed to `new_dim`.
+    num : int
+      If larger than 1 (default), the number of matrices to generate, stacked along a "matrices" dimension.
+    new_dim : str
+      Name of the new "prime" dimension, defaults to the same name as `crd` + "_prime".
+
+    Returns
+    -------
+    xr.DataArray
+      Float, NxN if num = 1, numxNxN otherwise, where N is the length of crd.
```

Comment:
```suggestion
    crd : xr.DataArray
        1D coordinate DataArray along which the rotation occurs.
        The output will be square with the same coordinate replicated,
        the second renamed to `new_dim`.
    num : int
        If larger than 1 (default), the number of matrices to generate, stacked along a "matrices" dimension.
    new_dim : str
        Name of the new "prime" dimension, defaults to the same name as `crd` + "_prime".

    Returns
    -------
    xr.DataArray
        Float, NxN if num = 1, numxNxN otherwise, where N is the length of crd.
```